[iOS] Adopt GCMouseInput.mouseMovedHandler to generate mouse move events when pointer lock is engaged

aprotyas opened the pull request:
#### 5b0cc9d273ea2ed0bf029e33c37f56c81d9a0819
<pre>
[iOS] Adopt GCMouseInput.mouseMovedHandler to generate mouse move events when pointer lock is engaged
<a href="https://bugs.webkit.org/show_bug.cgi?id=297305">https://bugs.webkit.org/show_bug.cgi?id=297305</a>
<a href="https://rdar.apple.com/157599039">rdar://157599039</a>

Reviewed by Tim Horton.

When UIPointerLockState.isLocked==YES, UIHoverGestureRecognizer
instances do not get relative offsets for pointer movements because
their locations are not updated at all, c.f. <a href="https://rdar.apple.com/57987082">rdar://57987082</a>. As a
result, our existing WKMouseInteraction machinery does not lend well to
event generation when pointer lock is engaged.

In this patch, we teach WKMouseInteraction about pointer lock by hooking
into pointer lock lifecycle methods on WebPageProxy. We do so by
introducing the notion of &quot;platform&quot; specific pointer lock steps, where
macOS hides the cursor through CG SPI. On the other hand, iOS plumbs
pointer lock state to WKMouseInteraction which then establishes a
mechanism to receive movement updates from a mouse pointing device. We
accomplish this through the GCMouseMoved handler on the GCMouseInput
object associated with the current mouse. Once set up, this handler is
used to generate NativeWebMouseEvent instances of mousemove type with a
locked location (i.e. the last touch location) and a raw movement delta.

We make WKMouseInteraction register for GCMouseDidDisconnectNotification
as a exit hatch in case the current mouse device disconnects from the
system. When notifications are received, we plumb this data back to
PointerLockController through the mouse interaction delegate.

Finally, this patch adds test coverage for this integration through the
PointerLockTests.MouseDevice[Move|Disconnect] API tests. To facilitate
testing, we instantiate a fake GCMouse instance and manually send a
delta to the GCMouseInput machinery under it. We link the GameController
framework against TestWebKitAPI, too, to be able to access the GCMouse
class and the GC notification key.

* Source/WebKit/Platform/Logging.h:
Add a PointerLock logging channel.

* Source/WebKit/Shared/ios/NativeWebMouseEventIOS.mm:
(WebKit::NativeWebMouseEvent::NativeWebMouseEvent):
Plumb the unadjustment movement deltas as a straight mapping of the
regular movement deltas.

* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::beginPointerLockMouseTracking):
(WebKit::PageClient::endPointerLockMouseTracking):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestPointerLock):
(WebKit::WebPageProxy::didAllowPointerLock):
(WebKit::WebPageProxy::requestPointerUnlock):
(WebKit::WebPageProxy::webContentPointerLockProcess):
(WebKit::WebPageProxy::clearWebContentPointerLockProcess):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::beginPointerLockMouseTracking):
(WebKit::PageClientImpl::endPointerLockMouseTracking):

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView mouseInteractionDidLoseMouseDeviceDuringPointerLock:]):
WKMouseInteractionDelegate method to inform the web page that we can no
longer stay engaged in pointer lock.
(-[WKContentView _beginPointerLockMouseTracking]):
(-[WKContentView _endPointerLockMouseTracking]):

* Source/WebKit/UIProcess/ios/WKMouseInteraction.h:
* Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:
(PointerLockState::reset):
(-[WKMouseInteraction _resetCachedState]):
(-[WKMouseInteraction createMouseEventWithType:wasCancelled:]):
(-[WKMouseInteraction dealloc]):
(-[WKMouseInteraction beginPointerLockMouseTracking]):
(-[WKMouseInteraction endPointerLockMouseTracking]):
(-[WKMouseInteraction handleGameControllerMouseMove:deltaY:]):
(-[WKMouseInteraction _startObservingMouseNotifications]):
(-[WKMouseInteraction _stopObservingMouseNotifications]):
(-[WKMouseInteraction _mouseDidStopBeingCurrent:]):
(-[WKMouseInteraction _setupMouseMovedHandlerForMouse:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::platformLockPointer):
(WebKit::WebPageProxy::platformUnlockPointer):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::platformLockPointer):
(WebKit::WebPageProxy::platformUnlockPointer):
* Tools/TestWebKitAPI/Configurations/TestWebKitAPIBase.xcconfig:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
(-[PointerLockDelegate waitForPointerLockEngaged]):
(-[PointerLockDelegate waitForPointerLockLost]):
(-[PointerLockDelegate _webViewDidRequestPointerLock:completionHandler:]):
(-[PointerLockDelegate _webViewDidLosePointerLock:]):
(-[PointerLockDelegate userContentController:didReceiveScriptMessage:]):
((PointerLockTests, MouseDeviceMove)):
((PointerLockTests, MouseDeviceDisconnect)):

Canonical link: <a href="https://commits.webkit.org/298789@main">https://commits.webkit.org/298789@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ecee693fd0931dc1e5a99e618eeffc774b0ab5d1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116643 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36307 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26869 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122717 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67215 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/0acf2d61-4a3e-4e0e-947a-b72910eeae9e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118532 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37005 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44896 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88588 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43005 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/cace5dcb-6b0c-43d0-9f38-a6601c9404ee) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119592 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29513 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104638 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69054 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28577 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22745 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66384 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98893 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22903 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125853 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43542 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32711 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97258 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43906 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100840 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97051 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42372 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20299 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39504 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18627 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43428 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49023 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42894 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46234 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44600 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:
```diff
@@ -236,7 +276,11 @@ inline static String pointerType(UITouchType type)
-    WebCore::DoublePoint point { [self locationInView:self.view] };
+    auto point = [&] -> WebCore::DoublePoint {
+        if (_pointerLockState.isActive)
+            return _pointerLockState.lockedCursorPosition.value_or(CGPointMake(0, 0));
```
```suggestion
            return _pointerLockState.lockedCursorPosition.value_or(CGPointZero);
```

aprotyas replied:
Fixed

pxlcoder commented on Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:
```diff
@@ -385,6 +429,118 @@ - (CGPoint)locationInView:(UIView *)view
+    if (!_pointerLockState.isActive)
+        return;
+
+    WebCore::DoublePoint lockedPoint { _pointerLockState.lockedCursorPosition.value_or(CGPointMake(0, 0)) };
```
```suggestion
    WebCore::DoublePoint lockedPoint { _pointerLockState.lockedCursorPosition.value_or(CGPointZero) };
```

aprotyas replied:
Fixed

rr-codes commented on Source/WebCore/platform/gamepad/cocoa/GameControllerSoftLink.h:
```diff
@@ -33,6 +33,7 @@
 SOFT_LINK_FRAMEWORK_FOR_HEADER(WebCore, GameController)
 SOFT_LINK_CLASS_FOR_HEADER(WebCore, GCController)
 SOFT_LINK_CLASS_FOR_HEADER(WebCore, GCControllerButtonInput)
+SOFT_LINK_CLASS_FOR_HEADER(WebCore, GCMouse)
```
where is this used in WebCore?

aprotyas replied:
It isn't. It was used in TestWebKitAPI, so I figured I drop the softlinking and just link against GameController for that target.

hortont424 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -14530,10 +14530,7 @@ void WebPageProxy::didAllowPointerLock(CompletionHandler<void(bool)>&& completio
-    CGDisplayHideCursor(CGMainDisplayID());
-    CGAssociateMouseAndMouseCursorPosition(false);
-#endif
+    platformPointerLock();
```
platformLockPointer()?

aprotyas replied:
Fixed

hortont424 commented on Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:
```diff
@@ -385,6 +429,118 @@ - (CGPoint)locationInView:(UIView *)view
+        return;
+
+    [self _stopObservingMouseNotifications];
+    [[std::exchange(_pointerLockState.currentMouse, nil) mouseInput] setMouseMovedHandler:std::exchange(_pointerLockState.originalMouseMovedHandler, nil)];
```
Some more locals might make this line less surprising to read

aprotyas replied:
Fixed

hortont424 commented on Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:
```diff
@@ -385,6 +429,118 @@ - (CGPoint)locationInView:(UIView *)view
+
+    WebCore::DoublePoint lockedPoint { _pointerLockState.lockedCursorPosition.value_or(CGPointZero) };
+    WebKit::NativeWebMouseEvent mouseEvent {
+        WebKit::WebEventType::MouseMove,
```
these constructors :(

aprotyas replied:
:(

hortont424 commented on Tools/TestWebKitAPI/Configurations/TestWebKitAPIBase.xcconfig:
```diff
@@ -133,9 +133,15 @@ WK_WRITING_TOOLS_UI_LDFLAGS_maccatalyst_MACOS_SINCE_1500 = -weak_framework Writi
 WK_WRITING_TOOLS_UI_LDFLAGS_MACOS_SINCE_1500 = -weak_framework WritingToolsUI;
 
+WK_GAMECONTROLLER_LDFLAGS = $(WK_GAMECONTROLLER_LDFLAGS_$(WK_PLATFORM_NAME))
+WK_GAMECONTROLLER_LDFLAGS_iphoneos = -framework GameController
```
Didn't Richard not do this for a reason??

aprotyas replied:
It's okay to do this in TestWebKitAPI

aprotyas committed:
[iOS] Adopt GCMouseInput.mouseMovedHandler to generate mouse move events when pointer lock is engaged
https://bugs.webkit.org/show_bug.cgi?id=297305
rdar://157599039

Reviewed by Tim Horton.

When UIPointerLockState.isLocked==YES, UIHoverGestureRecognizer
instances do not get relative offsets for pointer movements because
their locations are not updated at all, c.f. rdar://57987082. As a
result, our existing WKMouseInteraction machinery does not lend well to
event generation when pointer lock is engaged.

In this patch, we teach WKMouseInteraction about pointer lock by hooking
into pointer lock lifecycle methods on WebPageProxy. We do so by
introducing the notion of "platform" specific pointer lock steps, where
macOS hides the cursor through CG SPI. On the other hand, iOS plumbs
pointer lock state to WKMouseInteraction which then establishes a
mechanism to receive movement updates from a mouse pointing device. We
accomplish this through the GCMouseMoved handler on the GCMouseInput
object associated with the current mouse. Once set up, this handler is
used to generate NativeWebMouseEvent instances of mousemove type with a
locked location (i.e. the last touch location) and a raw movement delta.

We make WKMouseInteraction register for GCMouseDidDisconnectNotification
as a exit hatch in case the current mouse device disconnects from the
system. When notifications are received, we plumb this data back to
PointerLockController through the mouse interaction delegate.

Finally, this patch adds test coverage for this integration through the
PointerLockTests.MouseDevice[Move|Disconnect] API tests. To facilitate
testing, we instantiate a fake GCMouse instance and manually send a
delta to the GCMouseInput machinery under it. We link the GameController
framework against TestWebKitAPI, too, to be able to access the GCMouse
class and the GC notification key.

* Source/WebKit/Platform/Logging.h:
Add a PointerLock logging channel.

* Source/WebKit/Shared/ios/NativeWebMouseEventIOS.mm:
(WebKit::NativeWebMouseEvent::NativeWebMouseEvent):
Plumb the unadjustment movement deltas as a straight mapping of the
regular movement deltas.

* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::beginPointerLockMouseTracking):
(WebKit::PageClient::endPointerLockMouseTracking):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestPointerLock):
(WebKit::WebPageProxy::didAllowPointerLock):
(WebKit::WebPageProxy::requestPointerUnlock):
(WebKit::WebPageProxy::webContentPointerLockProcess):
(WebKit::WebPageProxy::clearWebContentPointerLockProcess):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::beginPointerLockMouseTracking):
(WebKit::PageClientImpl::endPointerLockMouseTracking):

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView mouseInteractionDidLoseMouseDeviceDuringPointerLock:]):
WKMouseInteractionDelegate method to inform the web page that we can no
longer stay engaged in pointer lock.
(-[WKContentView _beginPointerLockMouseTracking]):
(-[WKContentView _endPointerLockMouseTracking]):

* Source/WebKit/UIProcess/ios/WKMouseInteraction.h:
* Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:
(PointerLockState::reset):
(-[WKMouseInteraction _resetCachedState]):
(-[WKMouseInteraction createMouseEventWithType:wasCancelled:]):
(-[WKMouseInteraction dealloc]):
(-[WKMouseInteraction beginPointerLockMouseTracking]):
(-[WKMouseInteraction endPointerLockMouseTracking]):
(-[WKMouseInteraction handleGameControllerMouseMove:deltaY:]):
(-[WKMouseInteraction _startObservingMouseNotifications]):
(-[WKMouseInteraction _stopObservingMouseNotifications]):
(-[WKMouseInteraction _mouseDidStopBeingCurrent:]):
(-[WKMouseInteraction _setupMouseMovedHandlerForMouse:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::platformLockPointer):
(WebKit::WebPageProxy::platformUnlockPointer):
* Source/WebKit/UIProcess/mac/WebPageProxyMac.mm:
(WebKit::WebPageProxy::platformLockPointer):
(WebKit::WebPageProxy::platformUnlockPointer):
* Tools/TestWebKitAPI/Configurations/TestWebKitAPIBase.xcconfig:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
(-[PointerLockDelegate waitForPointerLockEngaged]):
(-[PointerLockDelegate waitForPointerLockLost]):
(-[PointerLockDelegate _webViewDidRequestPointerLock:completionHandler:]):
(-[PointerLockDelegate _webViewDidLosePointerLock:]):
(-[PointerLockDelegate userContentController:didReceiveScriptMessage:]):
((PointerLockTests, MouseDeviceMove)):
((PointerLockTests, MouseDeviceDisconnect)):

Canonical link: https://commits.webkit.org/298789@main
